File: LeetCode/Easy/0141. Linked List Cycle.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

bool hasCycle(struct ListNode *head) {
    if(head == NULL || head->next == NULL) return false;
    struct ListNode *curr = head;
    
    while(curr) {
        if(curr->val == -1000000) return true;
        curr->val = -1000000;
        curr = curr->next;
    }
    return false;
}

File: LeetCode/Easy/0141. Linked List Cycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool

        approach O(1) space:
        tortoise and hare method
        """

        tortoise = hare = head

        while hare:
            hare = hare.next
            if hare:
                hare = hare.next
            else:
                return False
            tortoise = tortoise.next
            if tortoise == hare: return True
        
        return False
        
